### PR TITLE
[red-knot] Add support for string annotations

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR adds support for parsing and inferring types within string annotations.

### Implementation (attempt 1)

This is preserved in https://github.com/astral-sh/ruff/pull/14151/commits/6217f48924f8f3f8a3d28c4929fe5aaad4ad0a59.

The implementation here would separate the inference of string annotations in the deferred query. This requires the following:
* Two ways of evaluating the deferred definitions - lazily and eagerly. 
	* An eager evaluation occurs right outside the definition query which in this case would be in `binding_ty` and `declaration_ty`.
	* A lazy evaluation occurs on demand like using the `definition_expression_ty` to determine the function return type and class bases.
* The above point means that when trying to get the binding type for a variable in an annotated assignment, the definition query won't include the type. So, it'll require going through the deferred query to get the type.

This has the following limitations:
* Nested string annotations, although not necessarily a useful feature, is difficult to implement unless we convert the implementation in an infinite loop
* Partial string annotations require complex layout because inferring the types for stringified and non-stringified parts of the annotation are done in separate queries. This means we need to maintain additional information

### Implementation (attempt 2)

This is the final diff in this PR.

The implementation here does the complete inference of string annotation in the same definition query by maintaining certain state while trying to infer different parts of an expression and take decisions accordingly. These are:
* Allow names that are part of a string annotation to not exists in the symbol table. For example, in `x: "Foo"`, if the "Foo" symbol is not defined then it won't exists in the symbol table even though it's being used. This is an invariant which is being allowed only for symbols in a string annotation.
* Similarly, lookup name is updated to do the same and if the symbol doesn't exists, then it's not bounded.
* Store the final type of a string annotation on the string expression itself and not for any of the sub-expressions that are created after parsing. This is because those sub-expressions won't exists in the semantic index.

Design document: https://www.notion.so/astral-sh/String-Annotations-12148797e1ca801197a9f146641e5b71?pvs=4

Closes: #13796 

## Test Plan

* Add various test cases in our markdown framework
* Run `red_knot` on LibCST (contains a lot of string annotations, specifically https://github.com/Instagram/LibCST/blob/main/libcst/matchers/_matcher_base.py), FastAPI (good amount of annotated code including `typing.Literal`) and compare against the `main` branch output

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -329,6 +330,8 @@ pub(super) struct TypeInferenceBuilder<'db> {
     /// The type inference results
     types: TypeInference<'db>,
 
+    deferred_state: DeferredState,
+
```
Comment (dhruvmanila) [inline crates/red_knot_python_semantic/src/types/infer.rs:334]: TODO: add some documentation

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -667,9 +669,6 @@ impl<'db> TypeInferenceBuilder<'db> {
         match definition.kind(self.db) {
             DefinitionKind::Function(function) => self.infer_function_deferred(function.node()),
             DefinitionKind::Class(class) => self.infer_class_deferred(class.node()),
-            DefinitionKind::AnnotatedAssignment(_annotated_assignment) => {
-                // TODO self.infer_annotated_assignment_deferred(annotated_assignment.node());
-            }
```
Comment (dhruvmanila) [inline crates/red_knot_python_semantic/src/types/infer.rs:672]: TODO: add a comment on why annotated assignment are not deferred

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 3]
```diff
@@ -4442,6 +4546,37 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 }
 
+#[derive(Default, Debug, Clone, Copy)]
+enum DeferredState {
+    #[default]
+    None,
+    Deferred,
+    InStringAnnotation,
+}
```
Comment (dhruvmanila) [inline crates/red_knot_python_semantic/src/types/infer.rs:4555]: TODO: add documentation for the enum and it's variants

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 4]
```diff
@@ -1064,14 +1063,15 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+        );
     }
 
     fn infer_class_deferred(&mut self, class: &ast::StmtClassDef) {
-        if self.are_all_types_deferred() {
-            for base in class.bases() {
-                self.infer_expression(base);
-            }
+        for base in class.bases() {
+            self.infer_expression(base);
```
Comment (dhruvmanila) [inline crates/red_knot_python_semantic/src/types/infer.rs:1074]: The diff is to remove the `self.are_all_types_deferred()` check because we're already in a deferred query here.

**Comment by dhruvmanila** [comment]
Although there are a couple of TODOs that I want to address mostly around documentation, I'd appreciate some initial feedback for any obvious issues that I've missed.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 5]
```diff
@@ -439,15 +443,13 @@ impl<'db> TypeInferenceBuilder<'db> {
...
-                }
+        if !self.types.deferred.is_empty() {
+            let mut deferred_inference = Vec::with_capacity(self.types.deferred.len());
+            for definition in &self.types.deferred {
+                deferred_inference.push(infer_deferred_types(self.db, *definition));
+            }
+            for deferred in deferred_inference {
+                self.extend(deferred);
             }
         }
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types/infer.rs:454]: Is it possible that inferring the deferred types add new entries to `self.types.deferred`?

I would probably write this as
```suggestion
				let deferred = std::mem::take(&mut self.deferred);
				for definition in deferred {
					self.extend(infer_derferred_types(self.db, *definition);
				}
        
        assert!(self.deferred.is_empty(), "Inferring deferred types shouldn't add new types to infer"); 
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 6]
```diff
@@ -4050,39 +4066,124 @@ impl<'db> TypeInferenceBuilder<'db> {
...
 impl<'db> TypeInferenceBuilder<'db> {
-    fn infer_annotation_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+    /// Infer the type of an annotation expression with the given [`DeferredState`].
+    fn infer_annotation_expression(
+        &mut self,
+        annotation: &ast::Expr,
+        deferred_state: DeferredState,
+    ) -> Type<'db> {
+        let previous_deferred_state = self.deferred_state;
+        self.deferred_state = deferred_state;
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types/infer.rs:4076]:
```suggestion
        let previous_deferred_state = std::mem::replace(&mut self.deferred_state, deferred_state);
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 7]
```diff
@@ -387,7 +391,7 @@ impl<'db> TypeInferenceBuilder<'db> {
 
     /// Are we currently inferring deferred types?
     fn is_deferred(&self) -> bool {
-        matches!(self.region, InferenceRegion::Deferred(_))
+        matches!(self.region, InferenceRegion::Deferred(_)) || self.deferred_state.is_deferred()
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:394]: Not sure if I'm missing an issue here, but maybe inferring an `InferenceRegion::Deferred` should set `self.deferred_state` and then this check can be simplified?

You have a note above to add docs for `deferred_state`; I think how these two things relate (and why `InferenceRegion::Deferred` overrides `self.deferred_state`) is a good candidate for some clear docs.

File: crates/red_knot_python_semantic/resources/mdtest/annotations/string.md [hunk 8]
```diff
@@ -1,9 +1,191 @@
 # String annotations
 
+## Simple
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/annotations/string.md:3]: These tests are fabulous! Thorough, clear, and correct, the testing trifecta :)

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 5 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs:454]: I agree it's important we either enforce and check that inferring deferred doesn't add more deferred, or else have explicit handling in case it does.

**Review by carljm — COMMENTED** [review]
Didn't complete review yet, but I have to run, so submitting the few comments I have so far in case I don't get back to it tonight.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 2 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs:672]: I'm curious to read this comment!

Could we add some tests that annotations on annotated assignments in stub files and when `from __future__ import annotations` is active, are in fact deferred? (That is, can contain forward references.)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 9]
```diff
@@ -1622,16 +1623,19 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+        let mut annotation_ty = self.infer_annotation_expression(
+            annotation,
+            DeferredState::from(self.are_all_types_deferred()),
+        );
 
         // Handle various singletons.
         if let Type::Instance(InstanceType { class }) = annotation_ty {
             if class.is_known(self.db, KnownClass::SpecialForm) {
-                if let Some(name_expr) = target.as_name_expr() {
+                if let Some(name_expr) = assignment.target.as_name_expr() {
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:1638]: Why the change to use `assignment.target` here instead of the `target` we unpacked above? Shouldn't they be the same?

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 10]
```diff
@@ -1644,7 +1648,7 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
         }
 
-        if let Some(value) = value {
+        if let Some(value) = assignment.value.as_deref() {
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:1651]: Why the change to use `assignment.value` here and no longer unpack `value` above?

File: crates/red_knot_python_semantic/src/types/string_annotation.rs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                Ok(parsed) => return Ok(parsed),
+                Err(parse_error) => diagnostics.add(
+                    string_literal.into(),
+                    "forward-annotation-syntax-error",
+                    format_args!("Syntax error in forward annotation: {}", parse_error.error),
+                ),
+            }
+        } else {
+            // The raw contents of the string doesn't match the parsed content. This could be the
+            // case for annotations that contain escaped quotes.
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/string_annotation.rs:60]:
```suggestion
            // case for annotations that contain escape sequences.
```

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 7 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs:394]: Ok after reading the whole diff more carefully, I think I understand this now, but yeah it would be great to document it.

**Review by carljm — APPROVED** [review]
This is excellent, great work! It turned out way less complicated than I thought it would be. I love that with a few small changes we can sidestep the whole issue of AST IDs and just store the resulting type on the string node. I think the result of this is that we won't be able to have hover types for individual parts of a string annotation, just the whole annotation, but personally I think that's fine; it is in fact just one expression, after all.

**Comment by dhruvmanila** [comment]
> Do we know if pyright supports hovering for stringified type annotations?

Yes, Pyright does support hover, goto definitions, references, etc. in string annotations.

Looking a bit deeper into Pyright, I think the reason they're able to do this is because they parse string annotations during the parsing stage and store it on the AST. But, the thing that I missed, and is not in the design document, is that the Pyright parser also takes into account `typing.Literal` and `typing.Annotated` with a limitation that it won't resolve the import and only consider it if it's exactly being imported from `typing` or `typing_extensions` (https://github.com/microsoft/pyright/blob/294b7afd2eaf23b0586bcc8563571bdff0c0d0a6/packages/pyright-internal/src/parser/parser.ts#L3610-L3621).

> Do we know how we would extend this design to support resolving types of inner expressions?

I and Micha talked about this in our 1:1 today and I think it's fine to move forward with this implementation today and visit it at a later stage. We might have to spend some additional time in figuring out the LSP part though. It might also be useful to understand the user impact of this feature if and when this needs to be implemented to validate the time investment. Some of the ideas for the implementation would be (a) updating the parser / AST to accommodate the change (b) semantic index that's specific to string annotation and an additional layer that connects the semantic index from the file with these specific ones.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 5 above)
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:454]: Yeah, makes sense, thanks

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 9 above)
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:1638]: Sorry, this is just a leftover from my initial attempt, will revert

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 10 above)
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:1651]: Same as above

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 2 above)
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:672]: Added the test cases in `assignment/annotations.md`

**Comment by carljm** [comment]
Yes, I think (b) describes how I'd envisioned we could tackle this, if/when we need to. I think it might also be possible to do something even simpler that doesn't do a full semantic index of the AST from the stringified annotation, but just adds a new mechanism for attaching a type directly to a text range instead of a node?
